### PR TITLE
Add Google Cloud DNS support

### PR DESCRIPTION
**Pull request by kamushadenes**
This adds Google Cloud DNS support.

User should create a Service Account and mount the JSON key file inside the container, specifying the path using the GOOGLE_APPLICATION_CREDENTIALS environment variable. User should also provide Google's Project ID, aka the GCP Environment, using the GOOGLEDNS_PROJECT_ID environment variable.

# Example Docker Compose

```
version: '2'
volumes:
  external-dns:
    external: true
    driver: rancher-nfs
services:
  google-dns-updater:
    image: rancher/external-dns
    environment:
      NAME_TEMPLATE: '%{{service_name}}.%{{stack_name}}.%{{environment_name}}'
      ROOT_DOMAIN: my-cool-company.com
      GOOGLEDNS_PROJECT_ID: company-123456
      GOOGLEDNS_MAX_RETRIES: '3'
      TTL: '5'
      GOOGLE_APPLICATION_CREDENTIALS: /external-dns/google/creds.json
    stdin_open: true
    volumes:
    - external-dns:/external-dns
    tty: true
    command:
    - -provider=googledns
    labels:
      io.rancher.container.agent.role: external-dns
      io.rancher.container.create_agent: 'true'
      io.rancher.container.pull_image: always
```

Tested in production with no issues so far.

### CODE REVIEW

**Comment by kamushadenes** [comment]
https://github.com/rancher/rancher/issues/9004

File: providers/googledns/googledns.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+//      credentials from the metadata server.
+//      (In this final case any provided scopes are ignored.)
+
+func (r *GoogleDNSProvider) Init(rootDomainName string) error {
+	r.limiter = ratelimit.NewBucketWithRate(5.0, 1)
+
+	if envVal := os.Getenv("GOOGLEDNS_MAX_RETRIES"); envVal != "" {
+		i, err := strconv.Atoi(envVal)
+		if err == nil {
+			googlednsMaxRetries = i
```
Comment (alena1108) [inline providers/googledns/googledns.go]: doesn't seem like this parameter is being used?

File: providers/googledns/googledns.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	r.limiter.Wait(1)
+
+	zones, err := r.client.ManagedZones.List(projectId).Do()
+
+	if len(zones.ManagedZones) == 0 {
+		return fmt.Errorf("Hosted zone for '%s' not found, got empty list", rootDomainName)
+	}
+
+	for _, zone := range zones.ManagedZones {
+		fmt.Printf("Found Zone: %v (%v)\n", zone.Name, zone.DnsName)
```
Comment (alena1108) [inline providers/googledns/googledns.go]: I would rather not print this intermediate info. You can log it in Debug. Also I suggest you use logrus for logging

File: providers/googledns/googledns.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+			if rrSet.Name == record.Fqdn {
+				recs = append(recs, rrSet)
+			}
+		}
+
+		if action == "UPDATE" {
+			change.Additions = records
+		}
+		change.Deletions = recs
```
Comment (alena1108) [inline providers/googledns/googledns.go]: for action UPDATE, do we also populate change.Deletions?

**Comment by ryangood** [comment]
Any idea as to when this might get merged into a release? Been looking for this feature for awhile to solve some blockers for us.

**Comment by kamushadenes** [comment]
@alena1108 Thanks for your feedback! I'll be checking your comments in a few minutes. 

Can I expect a merge after that?

File: providers/googledns/googledns.go (hunk 3 above)
Comment (kamushadenes) [reply providers/googledns/googledns.go]: Yes. There is no update method on the API, so in the same batch we have to remove the existing ones and add the updated entries.

File: providers/googledns/googledns.go (hunk 1 above)
Comment (kamushadenes) [reply providers/googledns/googledns.go]: You're right, I'll remove it.

**Comment by kamushadenes** [comment]
Fixed. I also squashed the commits.

**Comment by alena1108** [comment]
Thank you for your contribution, but with the release of Rancher 2.0, development on v1.6 is only limited to critical bug fixes and security patches

**Pull request closed without merging**